File: red_black_trees/red_black_trees.py
import enum


class Color(enum.Enum):

    RED = 'red'
    BLACK = 'black'


class Element:

    __slots__ = ['key', 'left', 'right', 'parent',  'color']

    def __init__(self, key, color=Color.RED):
        self.key = key
        self.left = self.right = self.parent = None
        self.color = color

    def __repr__(self):
        return f"{self.key}({self.color.value})"


sentinel = None


def nil():
    global sentinel

    if not sentinel:
        sentinel = Element(-1, color=Color.BLACK)

    return sentinel


class Tree:

    __slots__ = ['root', 'size', 'height']

    def __init__(self):
        self.root = nil()
        self.size = self.height = 0

    def put(self, z):
        self.size += 1

        x, y = self.root, nil()
        while x != nil():
            y = x
            x = x.left if z.key < x.key else x.right

        z.parent = y
        z.left = z.right = nil()
        if y == nil():
            self.root = z
        else:
            if z.key < y.key:
                y.left = z
            else:
                y.right = z

        self.__fix_after_insertion(z)

        self.height = self.__get_height()

    def contains(self, key):
        p = self.root
        while p != nil() and p.key != key:
            p = p.left if key < p.key else p.right

        return p

    def successor(self, z):
        if z.right != nil():
            p = z.right
            while p.left != nil():
                p = p.left

            return p
        else:
            p, q = z, z.parent
            while q != nil() and p == q.right:
                p = q
                q = p.parent

            return q

    def remove(self, z):
        self.size -= 1

        y = z
        color_of_y = y.color
        if z.left == nil():
            x = z.right
            self.__transplant(z, z.right)
        elif z.right == nil():
            x = z.left
            self.__transplant(z, z.left)
        else:
            y = self.successor(z)
            color_of_y = y.color
            x = y.right
            if y != z.right:
                self.__transplant(y, y.right)
                y.right = z.right
                y.right.parent = y
            else:
                x.parent = y

            self.__transplant(z, y)
            y.left = z.left
            y.left.parent = y
            y.color = z.color

        if color_of_y == Color.BLACK:
            self.__fix_after_deletion(x)

        z = None

        self.height = self.__get_height()

    def print_tree(self):

        def print_tree_internal(r, space_size=0):
            if r.left != nil():
                print_tree_internal(r.left, space_size=space_size+1)

            print("    " * space_size, r)

            if r.right != nil():
                print_tree_internal(r.right, space_size=space_size+1)

        print(f"size = {self.size} / height = {self.height}")

        if self.root != nil():
            print_tree_internal(self.root)

    def __fix_after_insertion(self, z):
        while z.parent.color == Color.RED:
            if z.parent == z.parent.parent.left:
                y = z.parent.parent.right
                if y.color == Color.RED:
                    z.parent.color = Color.BLACK
                    z.parent.parent.color = Color.RED
                    y.color = Color.BLACK
                    z = z.parent.parent
                else:
                    if z == z.parent.right:
                        z = z.parent
                        self.__rotate_left(z)
                    z.parent.color = Color.BLACK
                    z.parent.parent.color = Color.RED
                    self.__rotate_right(z.parent.parent)
            else:
                y = z.parent.parent.left
                if y.color == Color.RED:
                    z.parent.color = Color.BLACK
                    z.parent.parent.color = Color.RED
                    y.color = Color.BLACK
                    z = z.parent.parent
                else:
                    if z == z.parent.left:
                        z = z.parent
                        self.__rotate_right(z)
                    z.parent.color = Color.BLACK
                    z.parent.parent.color = Color.RED
                    self.__rotate_left(z.parent.parent)

        self.root.color = Color.BLACK

    def __fix_after_deletion(self, x):
        while x != self.root and x.color == Color.BLACK :
            if x == x.parent.left:
                w = x.parent.right
                if w.color == Color.RED:
                    # case 1
                    x.parent.color = Color.RED
                    w.color = Color.BLACK
                    self.__rotate_left(x.parent)
                    w = x.parent.right
                if w.left.color == Color.BLACK and w.right.color == Color.BLACK:
                    # case 2
                    w.color = Color.RED
                    x = x.parent
                else:
                    if w.right.color == Color.BLACK:
                        # case 3
                        w.left.color = Color.BLACK
                        w.color = Color.RED
                        self.__rotate_right(w)
                        w = x.parent.right
                    # case 4d
                    w.color = x.parent.color
                    x.parent.color = Color.BLACK
                    w.right.color = Color.BLACK
                    self.__rotate_left(x.parent)
                    x = self.root
            else:
                w = x.parent.left
                if w.color == Color.RED:
                    # case 1
                    x.parent.color = Color.RED
                    w.color = Color.BLACK
                    self.__rotate_right(x.parent)
                    w = x.parent.left
                if w.right.color == Color.BLACK and w.left.color == Color.BLACK:
                    # case 2
                    w.color = Color.RED
                    x = x.parent
                else:
                    if w.left.color == Color.BLACK:
                        # case 3
                        w.right.color = Color.BLACK
                        w.color = Color.RED
                        self.__rotate_left(w)
                        w = x.parent.left
                    # case 4
                    w.color = x.parent.color
                    x.parent.color = Color.BLACK
                    w.left.color = Color.BLACK
                    self.__rotate_right(x.parent)
                    x = self.root

        x.color = Color.BLACK

    def __rotate_left(self, x):
        y = x.right

        x.right = y.left
        if y.left != nil():
            y.left.parent = x

        y.parent = x.parent
        if x.parent == nil():
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y

        y.left = x
        x.parent = y

    def __rotate_right(self, x):
        y = x.left

        x.left = y.right
        if y.right != nil():
            y.right.parent = x

        y.parent = x.parent
        if x.parent == nil():
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y

        y.right = x
        x.parent = y

    def __transplant(self, p, q):
        if p.parent == nil():
            self.root = q
        elif p == p.parent.left:
            p.parent.left = q
        else:
            p.parent.right = q

        q.parent = p.parent

    def __get_height(self):

        def get_height_internal(r):
            if r == nil():
                return -1

            height_of_left = get_height_internal(r.left)
            height_of_right = get_height_internal(r.right)

            if height_of_left < height_of_right:
                return height_of_right + 1
            else:
                return height_of_left + 1

        if not self.root:
            return 0

        return get_height_internal(self.root)


if __name__ == '__main__':
    t = Tree()

    while True:
        print("1:put 2:remove 3:print 4:successor > ", end='')
        op = int(input())

        if op == 1:
            print("input key > ", end='')
            key = int(input())

            t.put(Element(key))
        elif op == 2:
            print("input key > ", end='')
            key = int(input())

            z = t.contains(key)
            if z == nil():
                print(f"{key} is not found in the tree.")
            else:
                t.remove(z)
        elif op == 3:
            t.print_tree()
        elif op == 4:
            print("input key > ", end='')
            key = int(input())

            z = t.contains(key)
            if z == nil():
                print(f"{key} is not found in the tree.")
            else:
                succ = t.successor(z)
                if succ == nil():
                    print("successor is none.")
                else:
                    print(f"successor = {succ.key}")
        else:
            print("invalid operation")

File: red_black_trees/red_black_trees.go
package main

import (
	"fmt"
	"strings"
)

type Element struct {
	key    int
	left   *Element
	right  *Element
	parent *Element
	color  int // 0:red, 1:black
}

type Tree struct {
	size   int
	height int
	root   *Element
}

var sentinel *Element

func nilElement() *Element {
	if sentinel == nil {
		sentinel = &Element{-1, nil, nil, nil, 1}
	}

	return sentinel
}

func (t *Tree) put(z *Element) {
	t.size += 1

	var x *Element = t.root
	var y *Element = nilElement()
	for x != nilElement() {
		y = x
		if z.key < x.key {
			x = x.left
		} else {
			x = x.right
		}
	}

	z.parent = y
	z.left = nilElement()
	z.right = nilElement()
	if y == nilElement() {
		t.root = z
	} else {
		if z.key < y.key {
			y.left = z
		} else {
			y.right = z
		}
	}

	t.fixAfterInsertion(z)

	t.height = t.getHeight()
}

func (t *Tree) rotateLeft(x *Element) {
	y := x.right
	x.right = y.left
	if y.left != nilElement() {
		y.left.parent = x
	}

	y.parent = x.parent
	if x.parent == nilElement() {
		t.root = x
	} else if x == x.parent.left {
		x.parent.left = y
	} else {
		x.parent.right = y
	}

	y.left = x
	x.parent = y
}

func (t *Tree) rotateRight(x *Element) {
	y := x.left
	x.left = y.right
	if y.right != nilElement() {
		y.right.parent = x
	}

	y.parent = x.parent
	if x.parent == nilElement() {
		t.root = x
	} else if x == x.parent.left {
		x.parent.left = y
	} else {
		x.parent.right = y
	}

	y.right = x
	x.parent = y
}

func (t *Tree) fixAfterInsertion(z *Element) {
	for z.parent.color == 0 {
		if z.parent == z.parent.parent.left {
			y := z.parent.parent.right
			if y.color == 0 {
				// case 1
				z.parent.color = 1
				y.color = 1
				z.parent.parent.color = 0
				z = z.parent.parent
			} else {
				if z == z.parent.right {
					// case 2
					z = z.parent
					t.rotateLeft(z)
				}

				// case 3
				z.parent.color = 1
				z.parent.parent.color = 0
				t.rotateRight(z.parent.parent)
			}
		} else {
			y := z.parent.parent.left
			if y.color == 0 {
				// case 1
				z.parent.color = 1
				y.color = 1
				z.parent.parent.color = 0
				z = z.parent.parent
			} else {
				if z == z.parent.left {
					// case 2
					z = z.parent
					t.rotateRight(z)
				}

				// case 3
				z.parent.color = 1
				z.parent.parent.color = 0
				t.rotateLeft(z.parent.parent)
			}
		}
	}

	t.root.color = 1
}

func (t *Tree) contains(key int) *Element {
	x := t.root
	for x != nil && x.key != key {
		if key < x.key {
			x = x.left
		} else {
			x = x.right
		}
	}

	return x
}

func (t *Tree) remove(z *Element) {
	t.size -= 1

	if z.left == nil {
		t.transplant(z, z.right)
	} else if z.right == nil {
		t.transplant(z, z.left)
	} else {
		y := t.successor(z)
		if y != z.right {
			t.transplant(y, y.right)
			y.right = z.right
			y.right.parent = y
		}

		t.transplant(z, y)
		y.left = z.left
		y.left.parent = z.parent
	}

	z = nil

	t.height = t.getHeight()
}

func (t *Tree) successor(x *Element) *Element {
	if x.right != nil {
		p := x.right
		for p.left != nil {
			p = p.left
		}

		return p
	} else {
		p := x
		q := x.parent
		for q != nil && p == q.right {
			p = q
			q = p.parent
		}

		return q
	}
}

func (t *Tree) transplant(x, y *Element) {
	if x.parent == nil {
		t.root = y
	} else if x == x.parent.left {
		x.parent.left = y
	} else {
		x.parent.right = y
	}

	if y != nil {
		y.parent = x.parent
	}
}

func (t *Tree) printTree() {
	fmt.Printf("size = %d, height = %d\n", t.size, t.height)

	if t.root != nilElement() {
		t.printTreeInternal(t.root, 0)
	}
}

func (t *Tree) printTreeInternal(r *Element, spaceSize int) {
	if r.left != nilElement() {
		t.printTreeInternal(r.left, spaceSize+1)
	}

	fmt.Printf("%s", strings.Repeat("    ", spaceSize))
	fmt.Printf("%d(", r.key)
	if r.color == 0 {
		fmt.Print("RED)")
	} else {
		fmt.Print("BLACK)")
	}
	fmt.Println()

	if r.right != nilElement() {
		t.printTreeInternal(r.right, spaceSize+1)
	}
}

func (t *Tree) getHeight() int {
	if t.root == nilElement() {
		return 0
	}

	return t.getHeightInternal(t.root)
}

func (t *Tree) getHeightInternal(r *Element) int {
	if r == nilElement() {
		return -1
	}

	heightOfLeft := t.getHeightInternal(r.left)
	heightOfRight := t.getHeightInternal(r.right)
	if heightOfLeft > heightOfRight {
		return heightOfLeft + 1
	} else {
		return heightOfRight + 1
	}
}

func main() {
	t := Tree{0, 0, nilElement()}

	for {
		fmt.Print("1:put 2:remove 3:print 4:successor > ")
		var op int
		fmt.Scanf("%d", &op)

		switch op {
		case 1:
			fmt.Print("input key > ")
			var data int
			fmt.Scanf("%d", &data)

			z := Element{data, nil, nil, nil, 0}
			t.put(&z)
		case 2:
			fmt.Print("input key > ")
			var key int
			fmt.Scanf("%d", &key)

			z := t.contains(key)
			if z == nil {
				fmt.Printf("%d is not found in the tree.\n", key)
			} else {
				t.remove(z)
			}
		case 3:
			t.printTree()
		case 4:
			fmt.Print("input key > ")
			var key int
			fmt.Scanf("%d", &key)

			z := t.contains(key)
			if z == nil {
				fmt.Printf("%d is not found in the tree.\n", key)
			} else {
				succ := t.successor(z)
				if succ != nil {
					fmt.Printf("successor = %d\n", succ.key)
				} else {
					fmt.Println("successor is not found.")
				}
			}
		default:
			fmt.Println("invalid operation")
		}
	}
}
